advisor/src/main/kotlin/advisors/VulnerableCode.kt

```diff
@@ -0,0 +1,168 @@
+
+            components.chunked(BULK_FETCH_SIZE).forEach { component ->
+                val requestResults =
+                    service.getPackageVulnerabilities(VulnerableCodeService.PackagesWrapper(component))
+                packagesToVulnerabilityRefs += requestResults
+            }
+
+            val packagesToVulnerabilities = fetchVulnerabilityDetails(packagesToVulnerabilityRefs, service)
+
+            val endTime = Instant.now()
+
+            packages.mapNotNull { pkg ->
+                packagesToVulnerabilities[pkg.id.toPurl()]?.let { vulnerabilities ->
+                    pkg to listOf(
+                        AdvisorResult(
+                            vulnerabilities,
+                            AdvisorDetails(advisorName),
+                            AdvisorSummary(startTime, endTime)
+                        )
+                    )
+                }
+            }.toMap()
+        } catch (e: Exception) {
+            e.showStackTrace()
+
+            val now = Instant.now()
+            val failedResults = listOf(
+                AdvisorResult(
+                    vulnerabilities = emptyList(),
+                    advisor = AdvisorDetails(advisorName),
+                    summary = AdvisorSummary(
+                        startTime = now,
+                        endTime = now,
+                        issues = listOf(
+                            createAndLogIssue(
+                                source = advisorName,
+                                message = "Failed to retrieve security vulnerabilities from $advisorName: " +
+                                        e.collectMessagesAsString()
+                            )
+                        )
+                    )
+                )
+            )
+            packages.associateWith { failedResults }
+        }
+    }
+
+    /**
+     * Construct a map from package URLs to the vulnerabilities of this package based on a
+     * [map with vulnerability references][packagesToVulnerabilityRefs] using the specified [service]. This
+     * function looks up the details of the vulnerabilities the references point to.
+     */
+    private suspend fun fetchVulnerabilityDetails(
+        packagesToVulnerabilityRefs: MutableMap<String, VulnerableCodeService.Vulnerabilities>,
+        service: VulnerableCodeService
+    ): Map<String, List<Vulnerability>> {
+        val vulnerabilityIds = packagesToVulnerabilityRefs.values.flatMap { it.unresolvedVulnerabilities }
+            .map { it.url }
+        val vulnerabilityDetails = resolveVulnerabilities(service, vulnerabilityIds)
+        return packagesToVulnerabilityRefs
+            .filterNot { it.value.unresolvedVulnerabilities.isEmpty() }
+            .map { entry ->
+                val vulnerabilities = entry.value.unresolvedVulnerabilities.mapNotNull { vulnerabilityDetails[it.url] }
+                entry.key to vulnerabilities
+            }.toMap()
```

Comment:
Nit, this could be more compact:

            .mapValues { (_, refs) ->
                refs.unresolvedVulnerabilities.mapNotNull { vulnerabilityDetails[it.url] }
            }.toMap()